### PR TITLE
🔴 VLM: compile compatibility

### LINKED ISSUE #29891: LLaVA `torch.compile` implementation
### Feature request

As per #28981, LLaVA is planned to receive `torch.compile` support. Seeing to the fact that LLaVA is composed of a vision tower and a LLM, both of which can be separately compiled with `fullgraph=True` (after support has been added, which is not the case for Mistral), it seems much easier to compile both parts separately as well.

### Motivation

The `_merge_input_ids_with_image_features` function that connects the two parts is difficult to compile as PyTorch has yet to add support for many of the functions used that require dynamic input sizes, which are necessary here as the number of input image tokens is subject to change.

### Your contribution

I'd love to try submitting a PR if possible but I'm not sure what the best way to do so is given the current circumstances.

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

As per title, adds flags in VLMs when needed, removes test skips and makes sure VLMs are compile compatible. Also for BLIP models adds new cache format in OPT which is one of backbones. Now all official BLIP models can support static cache and thus compile

NOTE: 
- Tests with `-k compile_forward` and `-k static_` were run for all models and are passing
- Regarding executorch which I also checked, the model can be exported and run a forward pass. But the generation won't work and probably would need smth similar to what we do when exporting VLMs in ONNX. Still need to dig more into that in later PRs

----------------
How to run compile and export for VLMs:
```python
import requests
from PIL import Image

import torch
from transformers import AutoProcessor, LlavaForConditionalGeneration
from transformers.generation import GenerationConfig
from transformers.cache_utils import StaticCache
from transformers.integrations.executorch import (
    TorchExportableModuleWithStaticCache,
    convert_and_export_with_cache,
)

model_id = "llava-hf/llava-interleave-qwen-0.5b-hf"
model = LlavaForConditionalGeneration.from_pretrained(
    model_id, 
    torch_dtype="float16", 
    device_map="cuda:0",
)
processor = AutoProcessor.from_pretrained(model_id)

conversation = [
    {

      "role": "user",
      "content": [
          {"type": "text", "text": "What are these?"},
          {"type": "image"},
        ],
    },
]
prompt = processor.apply_chat_template(conversation, add_generation_prompt=True)
image_file = "http://images.cocodataset.org/val2017/000000039769.jpg"
raw_image = Image.open(requests.get(image_file, stream=True).raw)
inputs = processor(images=raw_image, text=prompt, return_tensors='pt').to(0, torch.float16)

# Run with static cache which compiles the forward in decoding phase for you
output = model.generate(**inputs, max_new_tokens=20, cache_implementation="static")
print(processor.decode(output[0][2:], skip_special_tokens=True))



# Try to export with `torch.export`. NOTE: TorchExportableModuleWithStaticCache is not ready for VLMs
# and as mentioned above, VLMs might need to export 3 different modules as in ONNX. One for text embedding,
# one for vision backbone and one for the LM backbone with simple decoding token-by-token
max_generation_length = 1000
model = LlavaForConditionalGeneration.from_pretrained(
    model_id,
    device_map="cuda:0",
    torch_dtype="float16",
    attn_implementation="sdpa",
    generation_config=GenerationConfig(
        use_cache=True,
        cache_implementation="static",
        max_length=max_generation_length,
        cache_config={
            "batch_size": 1,
            "max_cache_len": max_generation_length,
        },
    ),
)

# Adapted from `TorchExportableModuleWithStaticCache` with minor changes
class TorchExportableModuleWithStaticCacheForVLM(torch.nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model
        self.static_cache = StaticCache(
            config=self.model.config.get_text_config(),
            batch_size=self.model.generation_config.cache_config.batch_size,
            max_cache_len=self.model.generation_config.cache_config.max_cache_len,
            dtype=self.model.dtype,
            device=model.device,
        )
        self.is_causal = any(("CausalLM" in arch or "ConditionalGeneration" in arch) for arch in self.model.config.architectures)
        if self.is_causal:
            causal_mask = torch.triu(
                torch.full(
                    (
                        self.model.generation_config.cache_config.batch_size,
                        1,
                        self.static_cache.max_cache_len,
                        self.static_cache.max_cache_len
                    ),
                    fill_value=torch.finfo(self.model.dtype).min,
                    dtype=self.model.dtype,
                    device=model.device,
                )
            )
            self.register_buffer("mask", causal_mask, persistent=False)

    def forward(
        self,
        input_ids: torch.Tensor,
        cache_position: torch.Tensor,
        pixel_values: torch.Tensor,
    ):
        _, seqlen = input_ids.shape
        attn_mask = self.mask[:, :, cache_position, :] if self.is_causal else None
        outs = self.model(
            input_ids=input_ids,
            attention_mask=attn_mask,
            position_ids=cache_position.unsqueeze(0),
            pixel_values=pixel_values,
            cache_position=cache_position,
            past_key_values=self.static_cache,
            use_cache=True,
        )
        return outs.logits

cache_position = torch.arange(inputs.input_ids.shape[1], dtype=torch.long, device=model.device)
export_inputs = {"input_ids": inputs.input_ids, "cache_position": cache_position, "pixel_values": inputs.pixel_values}

with torch.no_grad():
    exported_program = torch.export.export(
        TorchExportableModuleWithStaticCacheForVLM(model),
        args=(),
        kwargs=export_inputs,
        strict=True,
    )

torch.export.save(exported_program, "exported_llava.pt2")
exported_program = torch.export.load("exported_llava.pt2")
out = exported_program.module().forward(
    input_ids=inputs.input_ids,
    pixel_values=inputs.pixel_values,
    cache_position=cache_position,
)
```
 ---------------------
## Benchmark on "llava-hf/llava-onevision-qwen2-7b-ov-hf" using the same script we use for LLMs + dummy image in inputs

image

Fixes https://github.com/huggingface/transformers/issues/29891

### CODE REVIEW

File: src/transformers/models/opt/modeling_opt.py [hunk 1]
```diff
@@ -148,106 +157,67 @@ def forward(
...
         # get query proj
         query_states = self.q_proj(hidden_states) * self.scaling
-        # get key, value proj
-        if is_cross_attention and past_key_value is not None:
+        query_states = query_states.view(bsz, -1, self.num_heads, self.head_dim).transpose(1, 2)
+
+        if past_key_value is not None:
+            if not isinstance(past_key_value, EncoderDecoderCache):
+                curr_past_key_value = past_key_value
+            else:
```
Comment (zucchini-nlp) [inline src/transformers/models/opt/modeling_opt.py:165]: I dont know why but OPT model works as decoder-only but the attention is written as cross-attention (not used anywhere in codebase). So we need to support somehow BC while using the new `DynamicCache`

As a workaround I simply added a check on cache instance. Another possibility is to accept and return only the correct cache (self or cross attn) but that means all encoder-decoder models will need a change thus breaking BC

File: src/transformers/models/llava/configuration_llava.py [hunk 2]
```diff
@@ -37,8 +37,6 @@ class LlavaConfig(PretrainedConfig):
             The config object or dictionary of the vision backbone.
         text_config (`Union[AutoConfig, dict]`, *optional*, defaults to `LlamaConfig`):
             The config object or dictionary of the text backbone.
-        ignore_index (`int`, *optional*, defaults to -100):
-            The ignore index for the loss function.
```
Comment (zucchini-nlp) [inline src/transformers/models/llava/configuration_llava.py:41]: I believe this can be removed as it is not used anymore when merging inputs. We can also deprecate properly, but i don't think anyone uses it

File: tests/generation/test_utils.py [hunk 3]
```diff
@@ -2061,11 +2065,17 @@ def test_generate_compile_model_forward(self):
...
+            half_batch_size = self.model_tester.batch_size // 2
+            inputs_dict_1 = {k: v[:half_batch_size, ...] for k, v in inputs_dict.items() if "head_mask" not in k}
+            inputs_dict_2 = {
+                k: v[half_batch_size : half_batch_size * 2, ...]
+                for k, v in inputs_dict.items()
+                if "head_mask" not in k
+            }
+            self.assertTrue(
+                inputs_dict_1[model_class.main_input_name].shape == inputs_dict_2[model_class.main_input_name].shape
+            )
```
Comment (zucchini-nlp) [inline tests/generation/test_utils.py:2078]: some models cannot generate only from `input_ids`, so we pass the whole dict except for the `head_mask` (anyway we're removing head mask soon )

File: tests/models/aria/test_modeling_aria.py [hunk 4]
```diff
@@ -83,14 +83,14 @@ def __init__(
             moe_intermediate_size=4,
             moe_num_experts=4,
             moe_topk=2,
-            num_attention_heads=20,
+            num_attention_heads=8,
```
Comment (zucchini-nlp) [inline tests/models/aria/test_modeling_aria.py:86]: Aria had hidden size of 32 and 20 heads, which caused by division problems in tests when inferring `head_dim`

**Comment by zucchini-nlp** [comment]
Ready for review failing test is flaky otherwise everything is passing on my end, including slow test for compile/StaticCache

**Comment by gante** [comment]
@zucchini-nlp 

In the PR header we can read

> all VLMs have dynamic control in prepare_inputs_for_generation and thus skip test_compile_forward which compiles the model for pre-fill phase. But the test for decoding stage compile is green therefore I'm leaving the flag as True

`test_compile_forward` was the old name for `test_generate_compile_model_forward` if I'm not mistaken, back when it also did end-to-end compilation tests. We no longer have the end-to-end compilation tests, so this part of the PR header is no longer accurate, correct?

File: src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py [hunk 5]
```diff
@@ -187,8 +187,8 @@ def _attn(self, query, key, value, attention_mask=None, head_mask=None):
         batch_size, num_attention_heads, query_length, attn_head_size = query.size()
         key_length = key.size(-2)
 
-        query = query.view(batch_size * num_attention_heads, query_length, attn_head_size)
-        key = key.view(batch_size * num_attention_heads, key_length, attn_head_size)
+        query = query.reshape(batch_size * num_attention_heads, query_length, attn_head_size)
+        key = key.reshape(batch_size * num_attention_heads, key_length, attn_head_size)
```
Comment (gante) [inline src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py:191]: Out of curiosity, what drove the change?

File: src/transformers/models/llava/configuration_llava.py (hunk 2 above)
Comment (gante) [reply src/transformers/models/llava/configuration_llava.py:41]: If it is not being used for a while and there are no issues related to it, it should be fine

(if a user sets it, it will be absorbed by `**kwargs`)

File: src/transformers/models/opt/modeling_opt.py (hunk 1 above)
Comment (gante) [reply src/transformers/models/opt/modeling_opt.py:165]: Very much copy-paste from somewhere else, see [this comment](https://github.com/huggingface/transformers/pull/17088#issuecomment-1117974165) 

IMO, we can make our maintenance easier and assume no encoder-decoder stuff :) But don't spend more time here, eventually this will be rewritten with `modular`

File: tests/models/aria/test_modeling_aria.py [hunk 6]
```diff
@@ -287,10 +287,18 @@ def test_generate_from_inputs_embeds_0_greedy(self):
     def test_generate_from_inputs_embeds_1_beam_search(self):
         pass
 
-    @unittest.skip(reason="Unsupported")
+    @unittest.skip(reason="Dynamic control flow due to MoE")
```
Comment (gante) [inline tests/models/aria/test_modeling_aria.py:290]: I've tagged Aria as not supporting the static cache a few days ago -- I think we can delete these compilation-related skips 🤗

File: tests/models/idefics/test_modeling_idefics.py [hunk 7]
```diff
@@ -780,6 +780,14 @@ def test_contrastive_generate_low_memory(self):
     def test_custom_4d_attention_mask(self):
         pass
 
+    @unittest.skip(reason="IDEFICS cannot compile due to dynamic control flow when checking inputs")
```
Comment (gante) [inline tests/models/idefics/test_modeling_idefics.py:783]: like aria, this was fixed 2 days ago -- the manual skips can be deleted 🤗

**Review by gante — APPROVED** [review]
LGTM, thank you for all the work on VLMs 💛

**Comment by gante** [comment]
The failing test seems related to this PR :D

File: src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py (hunk 5 above)
Comment (zucchini-nlp) [reply src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py:191]: They were not contiguous when compiling, probably it wasn't noticed because model is super low usage.

**Review by ArthurZucker — COMMENTED** [review]
Very nice but 🔴 there are a few breaking changes so let's be careful! 
And do you have some benches / perf imporvements to share (making sure reduce overhead is working etc_

File: src/transformers/models/llava/modeling_llava.py [hunk 8]
```diff
@@ -498,12 +418,13 @@ def forward(
...
             )
 
-            n_image_tokens = (input_ids == self.config.image_token_index).sum().item()
-            n_image_features = image_features.shape[0] * image_features.shape[1]
-            if n_image_tokens != n_image_features:
-                raise ValueError(
-                    f"Image features and image tokens do not match: tokens: {n_image_tokens}, features {n_image_features}"
-                )
+            if not is_torchdynamo_compiling():
+                n_image_tokens = (input_ids == self.config.image_token_index).sum().item()
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/llava/modeling_llava.py:422]: sure, in paliagemma we do with `numel()` as another way to compare. Will change it and leave under `not is_compiling` condition

**Review by ArthurZucker — APPROVED** [review]

**Comment by zucchini-nlp** [comment]
image


Soooo, here is the correct eval with llava-ov-7b. One thing to note is that VLMs will not benefit from torch when the context length  is very high, like in case of videos or high-res images. That's the reason vanilla llava got speed up from first try, while llava-ov needed a few runs to notice how many input tokens we had

I will make fixup and merge this, because VLMs with less tokens per image get speedups
